src/derivkit/utils/numerics.py

```diff
@@ -56,3 +66,242 @@ def relative_error(a: np.ndarray, b: np.ndarray) -> float:
     b = np.asarray(b, dtype=float)
     denom = np.maximum(1.0, np.maximum(np.abs(a), np.abs(b)))
     return float(np.max(np.abs(a - b) / denom))
+
+
+def evaluate_logprior(
+    theta: ArrayLike,
+    logprior: Callable[[NDArray[np.floating]], float] | None,
+) -> float:
+    """Evaluates a log-prior with robust handling of invalid values.
+
+    If no prior is provided, this function represents an improper flat prior
+    and returns zero. If a prior is provided, its value is interpreted as a
+    log-density defined up to an additive constant. Any non-finite value
+    (e.g., ``-inf`` or ``nan``) is treated as zero prior probability.
+
+    Args:
+        theta: Parameter vector at which to evaluate the prior.
+        logprior: Callable returning the log-prior density, or ``None`` to
+            indicate an uninformative (flat) prior.
+
+    Returns:
+        The log-prior value at ``theta``, or ``-inf`` if the prior assigns
+        zero probability to that point.
+    """
+    if logprior is None:
+        return 0.0
+    v = float(logprior(np.asarray(theta)))
+    return v if np.isfinite(v) else -np.inf
+
+
+def in_bounds_checker(
+    theta: NDArray[np.floating],
+    bounds: Sequence[tuple[float | None, float | None]] | None,
+) -> bool:
+    """Tests whether parameters lie within optional per-parameter bounds.
+
+    Bounds are interpreted component-wise. For each parameter, either side may
+    be unbounded by setting that limit to ``None``. If no bounds are provided,
+    all points are considered valid.
+
+    Args:
+        theta: Parameter vector to test.
+        bounds: Optional sequence of ``(lower, upper)`` pairs, one per parameter.
+            Use ``None`` for an unconstrained lower or upper limit.
+
+    Returns:
+        True if all parameters satisfy their bounds (or if ``bounds`` is ``None``);
+        otherwise False.
+    """
+    if bounds is None:
+        return True
+    if len(bounds) != theta.size:
+        raise ValueError(f"bounds length {len(bounds)} != theta length {theta.size}")
+
+    for t, (lo, hi) in zip(theta, bounds):
+        if (lo is not None and t < lo) or (hi is not None and t > hi):
+            return False
+    return True
```

Comment:
This isn't Pythonic. A quick example using list comprehension:
```suggestion
    result = True
    if bounds is not None:
        if len(bounds) != theta.size:
            raise ValueError(f"bounds length {len(bounds)} != theta length {theta.size}")

        is_within_bounds = [True for x in theta for y in bounds if y is not None and x > y[0] and x < y[1]]
        all_in_bounds = not False in is_within_bounds
        
    return all_in_bounds
```
If you don't want to use list comprehension you could rewrite the for loop like
```
    for t, (lo, hi) in zip(theta, bounds):
        if (lo is not None and t < lo) or (hi is not None and t > hi):
            result = False
            break
```
This limits the branching.